### PR TITLE
OpenTelemetry Spring Boot example

### PR DESCRIPTION
**Pull request by aabmass**
## Description

Adds an example of a Spring Boot application instrumented with OpenTelemetry java agent. This example is a Spring version of [this golang sample](https://github.com/GoogleCloudPlatform/golang-samples/tree/main/opentelemetry/instrumentation) which was added in https://github.com/GoogleCloudPlatform/golang-samples/pull/3312.

Everything can be run with docker compose so it can easily be run in Cloud Shell. Additional details in the README

## Checklist

- [ ] I have followed [Sample Format Guide](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md)
- [ ] `pom.xml` parent set to latest `shared-configuration`
- [x] Appropriate changes to README are included in PR
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [x] **Tests** pass:   `mvn clean verify` **required** (actually using gradle)
- [x] **Lint**  passes: `mvn -P lint checkstyle:check` **required** (actually using gradle with spotless plugin)
- [ ] **Static Analysis**:  `mvn -P lint clean compile pmd:cpd-check spotbugs:check` **advisory only**
- [x] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample 
- [x] Please **merge** this PR for me once it is approved

## Permissions

This sample writes to Cloud Logging, Cloud Monitoring, and Cloud Trace. Grant the following roles to run the example:
- `roles/logging.logWriter` – see https://cloud.google.com/logging/docs/access-control#permissions_and_roles
- `roles/monitoring.metricWriter` – see https://cloud.google.com/monitoring/access-control#predefined_roles
- `roles/cloudtrace.agent` – see https://cloud.google.com/trace/docs/iam#trace-roles

### CODE REVIEW

**Review by minherz — CHANGES_REQUESTED** [review]
There are a couple of questions that we need to clarify before adding this code sample to the repo:

1) It is built only with gradle. Would it be possible to support both maven and gradle configurations?
2) Please provide tests for this code sample. It cannot be added "as is" because of its complexity and inability to maintain its integrity
3) I see that the region tags aren't referenced anywhere in documentation. This repo hosts code samples that are used in the documentation. What are the plans to use this code sample?

**Comment by aabmass** [comment]
Thanks for the review @minherz!

> * It is built only with gradle. Would it be possible to support both maven and gradle configurations?

Yes I think I could manage this. Is a usability concern or just for the tooling in this repo?

> * Please provide tests for this code sample. It cannot be added "as is" because of its complexity and inability to maintain its integrity

Do you mean integration tests as described in https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md#testing, or just unit tests?

> I see that the region tags aren't referenced anywhere in documentation. This repo hosts code samples that are used in the documentation. What are the plans to use this code sample?

I am working with Pamela and Mary on some docs which will be added under https://cloud.google.com/stackdriver/docs. I'll ping you some more details on chat

**Comment by aabmass** [comment]
@minherz I updated the sample to use maven and the parent POM.

I would appreciate your input on the testing story here. The Spring Boot setup has a basic smoke test included. Doing an integration test would involve running docker compose and verifying the telemetry output. I'm not sure how to do that based on the existing tests I see in this repo.

@punya @dashpole @psx95 would appreciate your input on OTel setup and Spring usage.

**Review by jsuereth — APPROVED** [review]
Overall Spring and OpenTelemetry conventions LGTM.

- Agree we should see what kind of tests we can produce.
- Downloading the Agent JAR via Docker file may need to be better called out.  It's unlikely dependabot will find this hook, however Maven doesn't really give you a good option (e.g. via Maven-jib-plugin) to create a docker image that uses its dependency resolution.

**Comment by aabmass** [comment]
> * Downloading the Agent JAR via Docker file may need to be better called out.  It's unlikely dependabot will find this hook, however Maven doesn't really give you a good option (e.g. via Maven-jib-plugin) to create a docker image that uses its dependency resolution.

Good point, I'll add some region tags and include it in the docs. Is there some way I can use maven to download the javaagent jar without including it in the app jar, and then manually copy it into the container?

**Review by dashpole — APPROVED** [review]

**Comment by aabmass** [comment]
Putting this back as draft for now until I:
- Downgrade to Spring Boot 2 for Java 8/11
- Update region tags
- Add additional tests

File: opentelemetry/spring-boot-instrumentation/src/test/java/com/example/demo/DockerComposeTestsIT.java
Comment (minherz) [inline opentelemetry/spring-boot-instrumentation/src/test/java/com/example/demo/DockerComposeTestsIT.java:52]: nit: is it possible to define the timeout for the test? can we decrease 1 minute run of the test?

File: opentelemetry/spring-boot-instrumentation/src/test/java/com/example/demo/DockerComposeTestsIT.java
Comment (minherz) [inline opentelemetry/spring-boot-instrumentation/src/test/java/com/example/demo/DockerComposeTestsIT.java:55]: nit: should the host to be 127.0.0.1 or another "fixed" value? The `8888` looks like the wrong default value for the host.

**Review by minherz — APPROVED** [review]
thank you for resolving the request about IT testing.
please, have a look into the comments i left.
there are several minor improvements that would be nice to make before merging the code sample.

the main question about this code sample is the use of region tags. in my opinion, the region tags don't reflect the Java code that has to be done to implement OTel instrumentation because the work isn't done in Java code. I encourage to reconsider the placement of region tags.

File: opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml
Comment (aabmass) [reply opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml:21]: It is "/var/log/app.log" which the application writes logs too. I can add a comment.

File: opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml
Comment (aabmass) [reply opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml:64]: No that is not the case. They are parsed into canonical OTel fields above which will get written to Cloud Logging correctly. They are removed here as they would just be duplicative

I realize this example is a bit backward in that it's basically round-tripping the logs through OTel collector back to Cloud Logging format. The goal here is to mimic the logging agent so that the application is portable. The Ops Agent would be a drop in replacement here.

File: opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml
Comment (aabmass) [reply opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml:82]: It's not necessary or recommended for this, the `googlecloud` exporter will handle splitting this into multiple payloads and requests to respect the limits you mentioned if needed.

File: opentelemetry/spring-boot-instrumentation/src/main/java/com/example/demo/SingleController.java
Comment (aabmass) [reply opentelemetry/spring-boot-instrumentation/src/main/java/com/example/demo/SingleController.java:33]: Do you have a recommendation for another location? The reasoning is to contextualize what the instrumentation is doing.

File: opentelemetry/spring-boot-instrumentation/docker-compose.yaml
Comment (aabmass) [reply opentelemetry/spring-boot-instrumentation/docker-compose.yaml]: removed

File: opentelemetry/spring-boot-instrumentation/README.md
Comment (aabmass) [reply opentelemetry/spring-boot-instrumentation/README.md]: Updated

File: opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml
Comment (aabmass) [reply opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml]: Updated

File: opentelemetry/spring-boot-instrumentation/src/test/java/com/example/demo/DockerComposeTestsIT.java
Comment (minherz) [reply opentelemetry/spring-boot-instrumentation/src/test/java/com/example/demo/DockerComposeTestsIT.java:55]: Sorry, I assumed that `getServiceHost` and `getServicePort` accepting a "default" value as a second argument.

File: opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml
Comment (minherz) [reply opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml:92]: Should we explain to a user that the resource will be identified as [global](https://cloud.google.com/monitoring/api/resources#tag_global)? Otherwise you should consider minimize the resource attributes to avoid confusion. If using `global` suites the example, follow the expected list of attributes for this type (i.e. `project_id`). You can place user attributes under `labels`.

File: opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml
Comment (minherz) [reply opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml:82]: I would highly recommend to place reasonable numbers here. They look random. And given that neither of them is near the allowed limits the claim for "more efficiently" does not held either.
I think that one or more of the other options will serve better as an example:

* add a comment with a link to the OTel batch configuration reference
* adjust the parameters to be closer to allowed limits i.e.
   ```yaml
   batch:
     send_batch_max_size: 10000
     timeout: 1s
   ```
* keep only `timeout` parameter
* update existing comment to explain that batch size can be controlled by time window or size

Comment (aabmass) [reply opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml:82]: @dashpole I copied these from the Go sample, were these param values chosen for any particular reason? I'm OK to go with the defaults except `timeout` here.

Re adjusting them to be similar to the limits, this is used for all three of metrics, logs, and traces pipelines below. I could make separate processors for each pipeline but I think it's overkill for the sample and will just complicate things.

File: opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml
Comment (aabmass) [reply opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml:92]: It won't be global, it will be `prometheus_target` with these labels

File: opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml
Comment (aabmass) [reply opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml:82]: For the purposes of the sample I agree, but a user might copy paste this

File: opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml
Comment (aabmass) [reply opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml]: Updated, PTAL

File: opentelemetry/spring-boot-instrumentation/src/main/java/com/example/demo/SingleController.java
Comment (minherz) [reply opentelemetry/spring-boot-instrumentation/src/main/java/com/example/demo/SingleController.java:33]: Since the "magic" is in the configuration files, shouldn't the region tag be placed in these files too?

Comment (aabmass) [reply opentelemetry/spring-boot-instrumentation/src/main/java/com/example/demo/SingleController.java:33]: I added region tags to the Dockerfile as well. I would like to keep this so the reader can understand what the sample app does inline in the docs

Comment (minherz) [reply opentelemetry/spring-boot-instrumentation/src/main/java/com/example/demo/SingleController.java:33]: I am unsure how this relates. Code snippets shown in the documentation demonstrate "what a developer has to do to implement a certain functionality in the specific programming language".
Putting region tag to the Java code "as is" will make developers to wander what exactly should be done.
This will be very confusing. The documentation should make it clear that no code changes in Java are required. And the changes should be done to container image configuration (in Dockerfile). However, it isn't part of the code sample.

File: opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml
Comment (jsuereth) [reply opentelemetry/spring-boot-instrumentation/otel-collector-config.yaml:87]: FYI - The batch size here should actually be optimised for *tracing* over both logging or metrics, as it's the most likely source of overload / need to batch.

I agree with @dashpole it can be removed, but I also think if you wanted to follow any kind of limit/restriction, it would be tracing limits.

Right now, TRACE quota [details](https://cloud.google.com/trace/docs/quotas) is not limited by this batch processor at all.   The set values of 500/1s lead to reasonable developer experience, but of course should be tuned based on actual collector overhead.

More importantly, batch values likely need to diverge by signal.  I agree with @dashpole on removing, but think this should be fine as-is.

File: opentelemetry/spring-boot-instrumentation/src/main/java/com/example/demo/SingleController.java
Comment (jsuereth) [reply opentelemetry/spring-boot-instrumentation/src/main/java/com/example/demo/SingleController.java:33]: If you think hollistically about the sample, and if you read any of our other language samples, you'll see that we do this to help avoid confusion.  Developers need to understand that their application *remains unchanged* when using autoinstrumentation.  Showing them this is a huge part of convincing them.

Specifically:

- The developer DOES need a starting application.  This callout will show that fact.
- We've seen in docs that developers need evidence that the code is not complicated or "doctored up" by tracing. If there's *one* thing this example does, it's help show spring developers that they don't need to write a lot of code to get tracing (one of the biggest concerns of tracing overall).

Comment (mtabasko) [reply opentelemetry/spring-boot-instrumentation/src/main/java/com/example/demo/SingleController.java:33]: Will we have a section in the docs that shows the Dockerfile snippet? It looks like that is being pulled under a separate region tag, so I'm assuming so.

If there really is nothing they need to do here, it's definitely a little weird -- both arguments here have merit, IMHO. It is a little weird to include a code snippet where nothing "additional" happens, but it's also helpful to know that you don't have to do anything to methods like this to get instrumentation.

A cursory glance suggests two possible approaches. to me:

One: Include the region tags here with a comment that tells them no code changes are needed for OTEL, ferex:

// [START opentelemetry_instrumentation_handle_single]
/** OpenTelemetry instrumentation requires no changes here. */
@GetMapping("/single")
  public String index() throws InterruptedException {
....

That way, we preserve the context for the user and make it clear that it's just that -- context -- no changes needed. We can even add a prefatory comment in the "Go" docs tab that says "No changes are needed" to reinforce.

Two: Skip the region tags here and just put a note in the "Go" docs tab that says something like "No changes are needed for Otel instrumentation". This doesn't give them any context of "where" such changes might have been needed.

Comment (minherz) [reply opentelemetry/spring-boot-instrumentation/src/main/java/com/example/demo/SingleController.java:33]: @mtabasko, I am happy with any option although showing a code with a comment that says smth like "this is just a code and it has nothing to do with the OTel instrumentation" is weird to me. If such technique is used in documentation of any product I am not aware of it.

Comment (minherz) [reply opentelemetry/spring-boot-instrumentation/src/main/java/com/example/demo/SingleController.java:33]: @jsuereth I do not think these arguments can be applied in this particular case. Google documentation shows code snippets to demonstrate _how to implement a particular functionality_. We don't argue about code samples but which part of them (aka code snippets) can or should be used in the documentation. If we were developing a tutorial then starting with a simple Spring Boot application and adding instrumentation could have merit.

In my understanding the targeting documentation is a simple howto article which does not require showing the code of the application in the doc.

However, I leave the final decision to @mtabasko. This PR can be merged once she makes the decision.

Comment (jsuereth) [reply opentelemetry/spring-boot-instrumentation/src/main/java/com/example/demo/SingleController.java:33]: So the issue is you assume the code DOESN'T matter at all.  It actually does.  OTEL Auto-instrumentation will infer certain characteristics based on the method name used in Spring.

Specifically, because this is the method `SingleController.index`, OTEL autoinstrumentation will use *that* name for one of the Spans it generates.  If the user were to change the method name or class name, a different Span name would be generated.

While the code doesn't need to change for autoinstrumentation, the code *is* needed to understand what autoinstrumentation is doing.  I'll defer to @mtabasko's judgement here as well.

Comment (mtabasko) [reply opentelemetry/spring-boot-instrumentation/src/main/java/com/example/demo/SingleController.java:33]: Well, then, by the powers vested in me by yinz, I hereby decree and declaim that we shall include the region tags to pull the snippet and add a comment to the tagged block to say something like "Otel instrumentation requires no changes here" (my first suggestion in my original comment). :)

@minherz This is a bit different from what we usually do in our code samples, but most of the samples we use (at least in Ops Mgmt) are related to our client libraries. This set of instrumentation docs we're working on is a bit different in emphasis, and I think that giving readers some context about what they DON'T have to do if they use OTel and Spring is appropriate here.

And if it turns out to be more confusing than helpful, we can always revise the sample and docs. But I know that if I were coming at this just from the docs, I would definitely find something like this useful as a wayfinding post in the journey.

**Comment by aabmass** [comment]
@minherz is there anything outstanding before this can be merged?

**Comment by minherz** [comment]
> @minherz is there anything outstanding before this can be merged?

I am OK with the content. Merge at your convenience.

**Pull request merged**
